core/src/main/java/jenkins/util/io/CompositeIOException.java

```diff
@@ -30,19 +30,52 @@
 import edu.umd.cs.findbugs.annotations.NonNull;
 import java.io.IOException;
 import java.io.UncheckedIOException;
+import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
 
 @Restricted(NoExternalUse.class)
 public class CompositeIOException extends IOException {
+    private static final long serialVersionUID = 121943141387608148L;
+
+    /**
+     * The maximum number of exceptions that can be reported by a single
+     * {@code CompositeIOException}.
+     * <p>
+     * The number of exceptions is limited to avoid pathological cases where
+     * where a huge number of exceptions could lead to excessive memory usage.
+     * For example, if the number of exceptions was unlimited, a call to
+     * {@code Util.deleteRecursive} could fail with a
+     * {@code CompositeIOException} that contains an exception for every
+     * single file inside of the directory.
+     */
+    public static final int EXCEPTION_LIMIT = 10;
```

Comment:
Not sure about naming, maybe `EXCEPTIONS_LIMIT` or `MAX_REPORTED_EXCEPTIONS` would be better. Same with the limit, maybe 15 or 20 would be a better default?

Reply:
Unclear why this is public. And if it's final, Groovy cannot change it.

Reply:
I think its because its used in tests

Reply:
Yes, it is public because it is used in tests. The class itself is annotated with `@Restricted(NoExternalUse.class)`. I made it final because it does not seem to me like something that would be useful for users to change, but like I mentioned in the PR description, if anyone thinks it is important to make it configurable I can make it non-final and allow it to be initialized via system property.